### ISSUE TITLE
Net 7 Kestrel windows service hangs after a period of time

### ISSUE THREAD

**Issue by sccrgoalie1** [dotnet/aspnetcore#46422]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We just update our projects to .NET 7 from .NET 5 and are experiencing hangs where Kestrel quits responding.  It works for a day or so and then the windows service needs to be restarted.  This was not an issue on .NET5.  I gathered a memory dump to see if I can figure out what is causing it to hang.  Here are some screenshots.  This was gathered using ProcDump.

This API is installed at many unique locations and several have reported the same issue that they need to restart the service daily.  For some locations, we use Azure relay.  So far, the locations that go through the relay have not had any issues.

![image](https://user-images.githubusercontent.com/2366790/216437164-f9a61114-5598-466e-b063-f9d2aebc87f6.png)

![image](https://user-images.githubusercontent.com/2366790/216437211-fa7a841b-7659-47b5-9659-89370fde674f.png)

### Expected Behavior

The API should respond while running

### Steps To Reproduce

Let our project run for a day

### .NET Version

7.0.100

**Comment by danmoseley** [comment]
Just curious does it repro on 6.0?

**Comment by sccrgoalie1** [comment]
> Just curious does it repro on 6.0?

Not sure, we skipped over 6.0 and went straight to 7.0.  We might be able to backtrack if that will help in diagnosing.

**Comment by davidfowl** [comment]
There was another issue like this with similar symptoms (hanging after a day or something).

**Comment by sccrgoalie1** [comment]
> There was another issue like this with similar symptoms (hanging after a day or something).

Yeah, it was this one https://github.com/dotnet/aspnetcore/issues/45215.  I reviewed that one and I don't think a solution was found, just not enough info to diagnose.

**Comment by davidfowl** [comment]
@sccrgoalie1 Can you run through all of the same steps?

**Comment by sccrgoalie1** [comment]
@davidfowl Absolutely, we ran through that issue extensively before we submitted our own.  Was there something specific that would be helpful?  I have the memory dump.  All the requests are http not https.  So far, it seems that the issue occurs after the API sits idle for a period of time.  Our users work banker's hours, so it works all day long, but when they come back the next morning Kestrel is no longer responding to any requests until we restart the windows service it's running in.

**Comment by adityamandaleeka** [comment]
@sccrgoalie1 Is it possible for you to share the dump with us (we'd understand if you cannot).

**Comment by adityamandaleeka** [comment]
Can you share more info about what exactly happens on the client and server when this happens?

For instance, does the client get a response back? Do the server logs show anything interesting? Perhaps you may get some interesting output if you enable more detailed logging on the server. See here for more info on configuring logging: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-7.0#configure-logging

**Comment by sccrgoalie1** [comment]
> @sccrgoalie1 Is it possible for you to share the dump with us (we'd understand if you cannot)?

@adityamandaleeka Sure, here is the memory dump.  

[Memory Dump](https://civicserviceorders3236.blob.core.windows.net/misc/miViewPointExtreme.CoreService.exe_230202_130106.dmp?sv=2021-10-04&st=2023-02-07T12%3A30%3A51Z&se=2023-02-09T12%3A30%3A00Z&sr=b&sp=r&sig=dI17Q9Xwcwq5QTU%2BR6H2qv13d5SSVLDPMB64hH2kQGM%3D)

**Comment by sccrgoalie1** [comment]
> Can you share more info about what exactly happens on the client and server when this happens?
> 
> For instance, does the client get a response back? Do the server logs show anything interesting? Perhaps you may get some interesting output if you enable more detailed logging on the server. See here for more info on configuring logging: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-7.0#configure-logging

The client does not get a response back, the requests just hang.  Unfortunately, we haven't found anything in the logs or event viewer yet.  We'll try increasing the detailed logging to see if we can get anything better.  The windows service on the server is still running but just not responding.

**Comment by adityamandaleeka** [comment]
@sccrgoalie1 Thanks for sharing the dump. I looked through it and didn't spot any immediate issues that jumped out. Just curious, did taking the dump cause the process to "unhang" or was it still in the stuck/unresponsive state?

**Comment by adityamandaleeka** [comment]
As a side note (maybe not related to the problem you're hitting), I see a bunch of threads doing things related to performance counters, which are running code from aspnet_perf.dll. Do you know what's causing that?

AFAICT aspnet_perf is an old ASP.NET framework component so I'm wondering what it's doing in the context of a ASP.NET Core app.

**Comment by sccrgoalie1** [comment]
> @sccrgoalie1 Thanks for sharing the dump. I looked through it and didn't spot any immediate issues that jumped out. Just curious, did taking the dump cause the process to "unhang" or was it still in the stuck/unresponsive state?

No, taking the dump did not unstick it.  At one of the sites having to restart daily we converted them to use  Microsoft.Azure.Relay.AspNetCore and since then they have been running for days without any issues.  I'm not sure at what level that package intertwines with Kestrel, but it seems to prevent the hang we are experiencing.

**Comment by sccrgoalie1** [comment]
> As a side note (maybe not related to the problem you're hitting), I see a bunch of threads doing things related to performance counters, which are running code from aspnet_perf.dll. Do you know what's causing that?
> 
> AFAICT aspnet_perf is an old ASP.NET framework component so I'm wondering what it's doing in the context of a ASP.NET Core app.

The only thing I can think of that might be doing that is Application Insights but we are using the NetCore version (Microsoft.ApplicationInsights.AspNetCore)

**Comment by heikkilamarko** [comment]
We have several .NET 7.0.2 applications running on Windows Server 2019. We run the apps as HashiCorp Nomad Jobs.

We noticed that HTTP requests to an application start to fail if we open the Windows "Resource Monitor" application and select the dotnet process running our app from the list (toggle check box on).
After selecting the application, it takes some time before HTTP requests, such as health checks start to fail. To make the app work again, restart is required.

You can also select more than one apps, and all the selected apps start to fail. App that are not selected work without problems.

**Comment by adityamandaleeka** [comment]
@heikkilamarko Thank you for reporting that. I was able to see the behavior you described even on a Windows 11 machine with an empty ASP.NET Core app.

Nothing obvious popped out under the debugger but cc @noahfalk @davmason in case this rings any bells. Presumably checking the box in Resource Monitor kicks off perfmon or something under the hood right?

**Comment by adityamandaleeka** [comment]
Hmm, AFAICT the app is still running and not hung, it's just not getting the incoming requests. @BrennanConroy pointed out that the browser is also not timing out so perhaps something is accepting and intercepting the communication...

**Comment by davmason** [comment]
It doesn't ring any bells for me. I am not sure exactly what resource monitor does under the hood

**Comment by heikkilamarko** [comment]
Glad that you were able to reproduce the problem. A small clarification. Some of our services are Node.js applications. They don't have this problem. The problem occurs only for dotnet apps.

**Comment by noahfalk** [comment]
@adityamandaleeka - No bells for me either I'm afraid. I suspect the right mental model for Resource Monitor and perfmon are that they are two independent apps that both make use of [perf counter APIs](https://learn.microsoft.com/en-us/windows/win32/perfctrs/consuming-counter-data).

**Comment by sccrgoalie1** [comment]
It seems possible that this is also what is causing the issues for our customers.  We noticed they all are running a software called Acronis.  It must be doing something similar to what Windows Resource monitor does to our ASP.NET Core app

**Comment by BrennanConroy** [comment]
We believe we narrowed down the issue to this change https://github.com/dotnet/runtime/pull/64834

This was verified by setting the environment variable `ComPlus_ThreadPool_UsePortableThreadPoolForIO` to `0` before running the app and being unable to repro the issue anymore.
@heikkilamarko and/or @sccrgoalie1 if you could also give that a try and let us know the results that would be great!

@kouvel do you have any ideas how the Windows Threadpool change could cause this issue?

**Issue summary:**
Running a Kestrel app and attaching Resource Monitor to the process will result in requests to the server hanging. It doesn't happen immediately, sometimes takes 5 minutes, sometimes 1 minute.
When using Wireshark to view TCP connections you can see that a connection is created to the port the server is listening on and keep-alives are sent, but Kestrel never sees the socket connection. I had a browser request finally close after 17.5 hours while running it overnight.
When setting `ComPlus_ThreadPool_UsePortableThreadPoolForIO=0` I have not seen the issue reproduce anymore.

**Comment by heikkilamarko** [comment]
Setting `ComPlus_ThreadPool_UsePortableThreadPoolForIO=0` seems to solve the problem.

**Comment by adityamandaleeka** [comment]
@heikkilamarko Glad you were able to confirm that. It should be a reasonable workaround for now while we investigate and fix the issue. Thank you again for the simple repro steps!

**Comment by kouvel** [comment]
It looks like for some reason, when Resource Monitor is attached, the IO for listening for connections is being associated with the thread it is issued on, even though the handle is attached to an IOCP. AFAIK that's not supposed to happen when the handle is attached to an IOCP. The IO happens to be issued on a thread pool worker thread, and when the thread times out and exits, the IO is cancelled with `ERROR_OPERATION_ABORTED`, and no more IO completions are received for that handle. It looks like worker threads in the portable thread pool implementation are [missing checking for pending IO](https://github.com/dotnet/runtime/blob/527278b8d78fbd14a78bc4f6bf40347453e40cc3/src/coreclr/vm/win32threadpool.cpp#L3281-L3288), that check prevents the thread from exiting and causing the IO to be canceled. It should be added, though I'm still not sure why that happens only when Resource Monitor is attached. I'll dig a bit deeper to see if I can find out what's happening.

**Comment by adityamandaleeka** [comment]
Thanks for the quick investigation @kouvel!

**Comment by adityamandaleeka** [comment]
Can you open an issue on runtime for this and link it from here?
